### DISCUSSION TITLE
Adding a suspension component halts dFdW

### DISCUSSION THREAD

**Discussion by arnalj** [mdolab/dafoam#559] [General]
Hello, I am trying to add a "suspension" component to the aerostructural scenario thus making a new "aerostructural+suspension" scenario. The way I've done this is by defining two new MDAO explicit components, a suspension component and a mesh movement component. The first takes in the drag, lift, and moment defined in `objFunc` of `daOptions` and outputs translation and rotation values. The second component takes in the translation and rotation, and moves the surface mesh accordingly. 

The forward problem works without problem, I am able to get a converged solution.

On the reverse pass however, the program gets stuck when calculating dFdW. It will print `Calculating dFdW using reverse-mode AD` and nothing will happen for hours and hours on end, the program doesn't crash.

I think the problem is associated with the derivatives of drag, lift, and moment. If I remove `d_outputs['drag']`, `d_outputs['lift']`, `d_outputs['moment']` from `compute_jacvec_product` in the suspension component, the program no longer halts. But of course the derivatives are no longer correct. Perhaps using "objective function" values as variables in the model causes problems?

My question is, do you have any insights on what could be causing this problem and some potential strategies to resolve it? I understand its hard to give specific advice without looking at code, but some general insights would be very much appreciated.

**Comment by friedenhe** [comment]
Can you elaborate on what are the inputs and outputs of the suspension component, and how you compute the outputs based on the inputs, and how you compute their derivatives?

**Reply by arnalj** [reply]
I attached the N2 diagram. [AeroStructSus.html.zip](https://github.com/mdolab/dafoam/files/13905603/AeroStructSus.html.zip)

The suspension component takes in point force coefficients (drag and lift) as well as the y-moment and outputs the movement in the z direction as well as the induced pitch angle. The suspension component also takes in spring coefficients as parameters as well as U0, A0, L0 to denormalize Cd Cl and CMY.

Given the forces, moments, spring coefficients, and geometry of the vehicle, we calculate the movement and rotation of the vehicle analytically by doing sum of F = 0, sum of M = 0. Note that the geometry is just the location of springs on the vehicle, these are hard coded in.

The derivatives are done by hand since the model is simple enough. I implemented the derivatives with `compute_jacvec_product`:

```
if mode == 'rev':
    if 'dz' in d_outputs:
        if 'CD' in d_inputs:
            d_inputs['CD'] += dXdf[0][0]*d_outputs['dz']
        if 'CL' in d_inputs:
            d_inputs['CL'] += dXdf[0][1]*d_outputs['dz']
        if 'CMY' in d_inputs:
            d_inputs['CMY']  += dXdf[0][2]*d_outputs['dz']
    if 'pitch_angle' in d_outputs:
        if 'CD' in d_inputs:
            d_inputs['CD'] += dXdf[1][0]*d_outputs['pitch_angle']
        if 'CL' in d_inputs:
            d_inputs['CL'] += dXdf[1][1]*d_outputs['pitch_angle']
        if 'CMY' in d_inputs:
            d_inputs['CMY']  += dXdf[1][2]*d_outputs['pitch_angle']
```

**Comment by friedenhe** [comment]
Not sure why the program hangs. But the code snippet you shared has an issue: You need to implement the matrix-vector product for all outputs and all inputs, so please also compute `d_inputs['z']` `d_inputs['L0']`, etc. If L0, U0, etc are constants, you should not define them as inputs. Also, I suggest you comment out everything in the compute_jacvec_product function, and then gradually uncomment each line and see which one breaks the code. For example, what is dXdf[0][0] and how it is computed?

**Comment by arnalj** [comment]
Ok thank you, I will remove the constants from inputs and just hard code them in for now. 

dXdf is a matrix that stores the hand derived derivatives. Its details were not important for the question, so I didn't include them in the question.

Thanks again Professor He.
